alacritty_terminal/src/url.rs

```diff
@@ -1,318 +1,40 @@
-// Copyright 2016 Joe Wilm, The Alacritty Project Contributors
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
+use crate::ansi::TermInfo;
+use crate::index::{Column, Linear, Point};
+use crate::term::Term;
+use std::ops::RangeInclusive;
 
-use unicode_width::UnicodeWidthChar;
-
-use crate::term::cell::{Cell, Flags};
-
-// See https://tools.ietf.org/html/rfc3987#page-13
-const URL_SEPARATOR_CHARS: [char; 10] = ['<', '>', '"', ' ', '{', '}', '|', '\\', '^', '`'];
-const URL_DENY_END_CHARS: [char; 7] = ['.', ',', ';', ':', '?', '!', '('];
-const URL_SCHEMES: [&str; 8] =
-    ["http://", "https://", "mailto:", "news:", "file://", "git://", "ssh://", "ftp://"];
-
-/// URL text and origin of the original click position.
-#[derive(Debug, PartialEq)]
 pub struct Url {
-    pub text: String,
-    pub origin: usize,
+    pub start: Point<usize>,
+    pub end: Point<usize>,
 }
 
-/// Parser for streaming inside-out detection of URLs.
-pub struct UrlParser {
-    state: String,
-    origin: usize,
-}
-
-impl UrlParser {
-    pub fn new() -> Self {
-        UrlParser { state: String::new(), origin: 0 }
-    }
-
-    /// Advance the parser one character to the left.
-    pub fn advance_left(&mut self, cell: &Cell) -> bool {
-        if cell.flags.contains(Flags::WIDE_CHAR_SPACER) {
-            self.origin += 1;
-            return false;
-        }
+impl Url {
+    pub fn new(start: Point<usize>, length: usize, num_cols: usize) -> Self {
+        let unwrapped_end_col = start.col.0 + length - 1;
+        let end_col = unwrapped_end_col % num_cols;
+        let end_line = start.line - unwrapped_end_col / num_cols;
 
-        if self.advance(cell.c, 0) {
-            true
-        } else {
-            self.origin += 1;
-            false
-        }
+        Url { end: Point::new(end_line, Column(end_col)), start }
     }
 
-    /// Advance the parser one character to the right.
-    pub fn advance_right(&mut self, cell: &Cell) -> bool {
-        if cell.flags.contains(Flags::WIDE_CHAR_SPACER) {
-            return false;
-        }
-
-        self.advance(cell.c, self.state.len())
+    pub fn contains(&self, point: impl Into<Point<usize>>) -> bool {
+        let point = point.into();
+        point.line <= self.start.line
+            && point.line >= self.end.line
+            && (point.line != self.start.line || point.col > self.start.col)
```

Comment:
You need inclusive check `point.col >= self.start.col` and `point.col <= self.end.col`

Reply:
Makes sense yeah, thanks!